pints/tests/test_residuals_diagnostics.py

```diff
@@ -0,0 +1,170 @@
+        self.values2 = self.model2.simulate(self.real_parameters2, self.times2)
+
+        # Add noise
+        self.noise2 = 0.05
+        self.values2 += np.random.normal(0, self.noise2, self.values2.shape)
+
+        # Create an object with links to the model and time series
+        self.problem2 = pints.MultiOutputProblem(
+            self.model2, self.times2, self.values2)
+
+        # Create a uniform prior over both the parameters and the new noise
+        # variable
+        self.log_prior2 = pints.UniformLogPrior([1, 1, 1, 1], [6, 6, 6, 6])
+        # Create a log likelihood
+        self.log_likelihood2 = pints.GaussianKnownSigmaLogLikelihood(
+            self.problem2, self.noise2)
+
+        # Create an un-normalised log-posterior (log-likelihood + log-prior)
+        self.log_posterior2 = pints.LogPosterior(
+            self.log_likelihood2, self.log_prior2)
+
+        # Run MCMC
+        self.x02 = [
+            self.real_parameters2 * 1.1,
+            self.real_parameters2 * 0.9,
+            self.real_parameters2 * 1.05
+        ]
+        mcmc = pints.MCMCController(self.log_posterior2, 3, self.x02)
+        mcmc.set_max_iterations(300)  # make it as small as possible
+        mcmc.set_log_to_screen(False)
+        self.samples2 = mcmc.run()
+
+    def test_calculate_residuals(self):
+        """
+        Test the calculate_residuals function.
+        """
+        # Test that it runs on an optimisation result
+        pints.residuals_diagnostics.calculate_residuals(
+            np.array([self.found_parameters1]), self.problem1)
+
+        # Test that it runs on MCMC samples without thinning
+        pints.residuals_diagnostics.calculate_residuals(self.samples2[0],
+                                                        self.problem2)
+
+        # Test that it runs on MCMC samples with thinning
+        pints.residuals_diagnostics.calculate_residuals(self.samples2[0],
+                                                        self.problem2,
+                                                        thinning=10)
+
+        # Check the message when the input is wrong dimension
+        self.assertRaisesRegexp(
+            ValueError,
+            r'\`parameters\` must be of shape \(n_samples\, n_parameters\)\.',
+            pints.residuals_diagnostics.calculate_residuals,
+            self.samples2,
+            self.model2
+        )
+
+        # Check the message when the thinning is invalid
+        self.assertRaisesRegexp(
+            ValueError,
+            r'Thinning rate must be \`None\` or an integer greater than '
+            r'zero\.',
+            pints.residuals_diagnostics.calculate_residuals,
+            self.samples2[0],
+            self.problem2,
+            thinning=0
+        )
+
+    def test_acorr(self):
+        """
+        Test the acorr function.
+        """
+        # Test that it runs without error
+        pints.residuals_diagnostics.acorr(np.random.normal(0, 1, 50), 11)
```

Comment:
Can we test this using series of a known autocorrelation structure? To check that the autocorrelations match the true values...

Reply:
Is something like this what you had in mind? A fixed series with precomputed autocorrelation is now hardcoded into the test

Reply:
Yep, perfect!